wasm2c/README.md

```diff
@@ -395,3 +455,104 @@ $ wat-desugar fac-flat.wat --fold -o fac-folded.wat
 
 The formatting is different and the variable and function names are gone, but
 the structure is the same.
+
+## Instantiate multiple instances of a module
+Since information about the execution context, such as memories, is encapsulated
+in the module instance structure, and a pointer to the structure is being passed through 
+function calls, multiple instances of the same module can be instantiated alongside
+one another.
+
+We can take a look at another version of the `main` function for a `rot13` example. By 
+declaring two sets of context information, two instances of `rot13` can be instantiated 
+in the same address space.
+
+```c
+#include "rot13.h"
+
+/* Define structure to hold the imports */
+struct Z_host_module_instance_t {
+  wasm_rt_memory_t s_memory;
+  char* s_input;
+};
+
+/* Accessor to access the memory member of the host */
+wasm_rt_memory_t* Z_hostZ_mem(struct Z_host_module_instance_t* instance) {
+  return &instance->s_memory;
+}
+
+/* Declare the implementations of the imports. */
+static u32 fill_buf(struct Z_host_module_instance_t* instance, u32 ptr, u32 size);
+static void buf_done(struct Z_host_module_instance_t* instance, u32 ptr, u32 size);
+
+/* Define host-provided functions under the names imported by the `rot13` instance */
+u32 Z_rot13Z_hostZ_fill_buf(struct Z_host_module_instance_t* instance,
```

Comment:
I would expect this to not have the `Z_rot13` prefix..   Isn't that prefix reserved for things defined in the rot32 module?   For things imported from the `host` module I would just expect the `Z_host` prefix.

Reply:
Welll.... our general theory had been that every public symbol (import or export) gets prefixed with the module prefix to keep things isolated. E.g. (in a world without "Z"), it seems scary if a WebAssembly module could import a function named `kill` from a module named `pthread`, and now it can call the real `pthread_kill(3)` function whenever it wants.

(This also gives the host a little more power to decide what it wants to export to every module type. E.g. maybe it wants to version its API by exporting one version of the `fill_buf` function to rot13_v1, but a different version of `fill_buf` to rot13_v2.)

Reply:
The downside/consequence is that the host does have to declare functions for each module it wants to be able to import something from it. If two modules (named "modulea" and "moduleb") each import a whole bunch of functions from "wasi_snapshot_preview1", then the host will have to arrange, probably via some preprocessor macro that sets this up, to declare functions named, e.g., `Z_moduleaZ_wasi_snapshot_preview1Z_fd_read` and also `Z_modulebZ_wasi_snapshot_preview1Z_fd_read`  that turn around and call the "real" WASI functions, in order for both modules to be allowed to call them.

Reply:
I think imports can be already considered prefixed/isloated because they contain the module name within them.  `Z_wasi_snapshot_preview1Z_fd_read` already contains that the `Z_wasi_snapshot_preview1` prefix.  If we want to more fine grained control, for example to be able to have two different version of `Z_wasi_snapshot_preview1Z_fd_read` in the same process, then I think we can address that as a followup and it will most likely involve using some kind of user level linking (e.g. structs of function pointers) rather than relying on the C symbol names to do the linking.

Reply:
Hmm, this still seems pretty scary to me. Maybe I'm off-base, so let me see if I can explain why. I put a proposed solution at the bottom that hopefully will satisfy your concern and also our fears...?

I hope I'm using the right language here. We see these three things as distinct:
1. the CWriter "module_prefix," a string controlled by the caller of wasm2c (if unspecified, the module's name if it has one, or else the filename)
2. the "mod:name" of an import, a string that the module producer/author controls
3. the "nm:name" of an import or export, also a string that the module producer/author controls

In this PR, the **exported** functions turn into function definitions, with the function named "module_prefix" + "nm:name". I think we're all on the same page on that one.

What we're talking about here is about the **imported** functions, which in the current PR become extern declarations referencing functions named with the form "module_prefix" + "mod:name" + "nm:name".

What I think you're suggesting is to rename them just "mod:name" + "nm:name" (removing the "module_prefix"). There's basically two reasons we're scared of this:

1. The "mod:name" and "nm:name" are both totally under the control of the module producer/author. If the module chooses to import a function named "kill" from a mod:name "pthread", then CWriter would emit an extern declaration for a function named "pthread_kill" (assuming no "Z" here), and then if linked into an executable, the WebAssembly module would be able to call that function (or "rt_sigreturn" or "timer_delete"...) with no intervention or explicit decision to grant that capability by the host/embedder. That doesn't seem right -- and so we think the import symbols all should require a prefix that's **not** under the control of the module author/producer.
2. Even if we prefixed all import and export symbols by some distinguished but constant prefix (e.g. "w2c_" + "mod:name" + "nm:name"), it's still scary because it means that two modules would be able to call one another in a possible signature-mismatched way. E.g. if one module exports a function named w2c_modname_banana with one type signature, and another module imports a function named w2c_modname_banana with a different type signature, the user will be allowed to link the respective .o files together without any warning message and will get garbage at runtime. And that seems to break the guarantees of WebAssembly.

-------------

**How about this as a proposed solution?**
- The exported functions are defined with names like "w2c_export_" + module_prefix + nm:name
- The imported functions are declared with names like "w2c_import_" + mod:name + nm:name
- As in the current PR, it remains the host's job to define a function that matches each import declaration. But unlike the current PR, these functions would no longer be specific to each module_prefix.

This preserves the two-level hierarchy (which I think you want) but also prevents modules from "automatically" linking with random libc functions or with the exports of another function (which is our big fear).

Reply:
> Hmm, this still seems pretty scary to me. Maybe I'm off-base, so let me see if I can explain why. I put a proposed solution at the bottom that hopefully will satisfy your concern and also our fears...?
> 
> I hope I'm using the right language here. We see these three things as distinct:
> 
> 1. the CWriter "module_prefix," a string controlled by the caller of wasm2c (if unspecified, the module's name if it has one, or else the filename)
> 2. the "mod:name" of an import, a string that the module producer/author controls
> 3. the "nm:name" of an import or export, also a string that the module producer/author controls
> 
> In this PR, the **exported** functions turn into function definitions, with the function named "module_prefix" + "nm:name". I think we're all on the same page on that one.
> 
> What we're talking about here is the **imported** functions, which in the current PR become extern declarations referencing functions named with the form "module_prefix" + "mod:name" + "nm:name".
> 
> What I think you're suggesting is to rename them just "mod:name" + "nm:name" (removing the "module_prefix"). There's basically two reasons we're scared of this:
> 
> 1. The "mod:name" and "nm:name" are both totally under the control of the module producer/author. If the module chooses to import a function named "kill" from a mod:name "pthread", then CWriter would emit an extern declaration for a function named "pthread_kill" (assuming no "Z" here), and then if linked into an executable, the WebAssembly module would be able to call that function (or "rt_sigreturn" or "timer_delete"...) with no intervention or explicit decision to grant that capability by the host/embedder. That doesn't seem right -- and so we think the import symbols all should require a prefix that's **not** under the control of the module author/producer.

Given that we do have the Z_ prefix on both the module name and the field name I don't think this is (currently) a valid concern.   How about we revisit this namespace issue if/when we remove the `Z_` managing we can revisit this question.  For now the I think the name mangling avoids any collisions worries with system symbols.

> 2. Even if we prefixed all import and export symbols by some distinguished but constant prefix (e.g. "w2c_" + "mod:name" + "nm:name"), it's still scary because it means that two modules would be able to call one another in a possible signature-mismatched way. E.g. if one module exports a function named w2c_modname_banana with some type signature, and another module imports a function named w2c_modname_banana with a different type signature, the user will be allowed to link the respective .o files together without any warning message and will get garbage at runtime (since the symbols have the same name after we removed the signature-mangling). And that seems to break the guarantees of WebAssembly.

This does sounds like a problem, but maybe not one that needs to be solved as part of this PR.  The problem sounds like "how can we link wasm2c modules together in a type-safe way?"   I'm not sure the native static linker is powerfull enough be used as way to link mouldes together so maybe we need to build some higher level linking solution where imports and exports are handling at a high level (rather than relying on the static linker to link things together).   Given this issue perhaps removing the signature from the symbol name in #1896 was actually a bad idea?   

> 
> **How about this as a proposed solution?**
> 
> * The exported functions are defined with names like "w2c_export_" + module_prefix + nm:name
> * The imported functions are declared with names like "w2c_import_" + mod:name + nm:name
> * As in the current PR, it remains the host's job to define a function that matches each import declaration. But unlike the current PR, these functions would no longer be specific to each module_prefix.
> 
> This preserves the two-level hierarchy (which I think you want) but also prevents modules from "automatically" linking with random libc functions or with the exports of another function (which is our big fear).

Sure that sounds reasonable.   This is still not powerful enough to express the use case where the host night want to export two different versions of the same function to different modules, but I think to achieve that (as I say above) we would need to switch to a runtime linking system that allows the host to hook up imports and exports dynamically rather than statically.

If possible, could all of this be differed to separate PR and keep this PR focused just on instancing?

Reply:
> If possible, could all of this be differed to separate PR and keep this PR focused just on instancing?

Okay -- just pushed a commit that does this. (Moves imports to just "mod:name" + "nm:name", with caveats above.) We can revisit the naming later, perhaps at the same time we remove "Z". Also updated wasm2c/README.md; the imported function names from the "host" module are now as you expected.

Re: structs of function pointers, I think for performance reasons we'd love to avoid a world where every call to an imported function has to go through a function pointer, but I admit I haven't actually tried to measure what the penalty would be there.

Reply:
Perhaps we need two different options for linking modules together:

1. Allow the host do set up imports dynamically allowing full flexibility at runtime
2. Have the static linker link everything together based on symbol names alone (i.e. matching up imports and exports symbolically).

Regarding the indirection .. don't we already use a pointer indirection for imports.   Looking at the spec test generated code for example I see:

```
extern void (*Z_spectestZ_print_i32)(u32); 
```

This is declaring a pointer to a function rather than a function so we already have that level of indirection there.  (Unless this change removes it?)

Reply:
Certainly seems like the linking issue (along with how type safety is enforced during linking) is something we should discuss further, outside of this issue.

Reply:
> Perhaps we need two different options for linking modules together:
> 
>     1. Allow the host do set up imports dynamically allowing full flexibility at runtime
> 
>     2. Have the static linker link everything together based on symbol names alone (i.e. matching up imports and exports symbolically).

Maybe there's a way to enforce type safety in both settings somehow, by using the fact that we have a single global data structure of unique func types (and we know where every exported func falls in that data structure). But yes, happy to discuss later outside this issue.

> Regarding the indirection .. don't we already use a pointer indirection for imports. Looking at the spec test generated code for example I see:
> 
> ```
> extern void (*Z_spectestZ_print_i32)(u32); 
> ```
> 
> This is declaring a pointer to a function rather than a function so we already have that level of indirection there. (Unless this change removes it?)

Yes, all those pointers end up going away in this PR, at least as it currently stands -- not just the function pointers, but also the table, memory, and global pointers. The table, memory, and globals stop being pointers because now they're just members of the Z_spectest_module_instance_t structure, and any module that imports from spectest takes a Z_spectest_module_instance_t* as an argument to its init function, and stores that pointer in its own module instance structure. And the functions stop being pointers because now they're just ordinary functions that take a Z_spectest_module_instance_t* as their first argument, provided by the importing module when calling the function.